ws-server/src/server.rs

```diff
@@ -179,18 +189,44 @@ async fn background_task(
 	}
 }
 
+#[derive(Debug, Clone)]
+enum Cors {
+	AllowAny,
+	AllowList(Arc<[String]>),
+}
+
+impl Cors {
+	fn verify_origin(&self, origin: Option<&[u8]>) -> Result<(), Error> {
+		if let (Cors::AllowList(list), Some(origin)) = (self, origin) {
+			if !list.iter().any(|o| o.as_bytes() == origin) {
+				let error = format!("Origin denied: {}", String::from_utf8_lossy(origin));
+				log::warn!("{}", error);
+				return Err(Error::Request(error));
+			}
+		}
```

Comment:
Do we want/need to handle "*" patterns as well in the allowed origins?

Reply:
yeah, I think so

Reply:
Yes we do.

Reply:
There are 3 possible states here:

1. There is no `Origin` header, which equates to the request being done either on the same domain (no cross-origin shenanigans), or the request isn't coming from a browser at all (in which case the client can spoof the header to whatever it wants). We always allow these.
2. The header has a `protocol://hostname` value, if so we check it against the list (unless we allow all origins).
3. The header has a `null` value, which can be explicitly allowed by the list, but is generally advised against.

There is no `*` origin.

Reply:
I think you explained this in a previous comment which makes sense to me now :)

My initial thinking was that CORS responses can contain "*" in allowed origins, so perhaps we need to handle something like this.

I did a little reading about WebSockets (since I didn't really know anything about the WebSocket upgrade protocol) and, indeed, there is no such thing as CORS really when talking about WS connections by the sounds of it, and so the origin checking here can essentially take whatever form we like.

Given that, I have no problem with not handling "*" in an allowed origin; it could potentially be a future enhancement (so that you can say configure this to allow eg any connections from "https://*.mydomain.com") but offhand I don't see why it needs to exist now.